File: neetcode-roadmap/arrays/004-Group-Anagrams-Medium/004-Group-Anagrams-Medium.cpp
// Problem link: https://leetcode.com/problems/group-anagrams/description/

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {

        vector<vector<string>> ans;
        std::unordered_map<string, vector<string>> sortMap;
        for(int i = 0; i < strs.size(); i++)
        {
            // current word we're on, sort it, if it exists in hashmap, append the unsorted word
            // to the values using the sorted word (key)
            // if it doesnt exist, create it

            string word = strs[i];
            string sortedKey = word;
            std::sort(sortedKey.begin(), sortedKey.end());
            if(sortMap.find(sortedKey) != sortMap.end())
            {
                sortMap[sortedKey].push_back(word);
            }
            else
            {
                sortMap[sortedKey] = {strs[i]};
            }
        }

        for(auto [key, value] : sortMap)
        {
            ans.push_back(value);
        }


        return ans;
        
    }
};

File: neetcode-roadmap/arrays/004-Group-Anagrams-Medium/004-Group-Anagrams-Medium.py
# Problem link: https://leetcode.com/problems/group-anagrams/description/

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        sortMap = {}
        ans = []
        
        for word in strs:
            sorted_key = ''.join(sorted(word)) # sorted(word) is a list of characters, so must join them
            if sorted_key not in sortMap:
                sortMap[sorted_key] = [word] # store word in a list (not just a str)
            else:
                sortMap[sorted_key].append(word)
        
        # group like this:
        # eat => a e t
        # tea => a e t
        # etc

        for k, v in sortMap.items():
            ans.append(v)

        return ans
        
